### DISCUSSION TITLE
understand memory leak from the flamegraph

### DISCUSSION THREAD

**Discussion by JackieJin1025** [bloomberg/memray#485] [Q&A]
Hi, I ran a script with memray, in the hope of finding the place of memory leak. Below is the snapshot of the memory usage from the flamegraph.  

<img width="1106" alt="image" src="https://github.com/bloomberg/memray/assets/8197215/f8512c0c-7b59-42b9-accb-418eb3080ef4">
There is a loop in the script that corresponds to the several peaks in the snapshot. As you can see, the memory slowly increases. Below is the code that executes the loop. I believe the line 72 simulation(....) should have a similar memory usage in each iteration. 
<img width="1191" alt="image" src="https://github.com/bloomberg/memray/assets/8197215/c986b24c-0d0e-4c2c-b588-cddf26743626">

The flamegraph does not seem to show the slow increase of memory over the loop. Is there a better way to find the memory leak?
<img width="1791" alt="image" src="https://github.com/bloomberg/memray/assets/8197215/8ef7c41a-8988-4797-bb03-8de93438f993">

**Comment by pablogsal** [comment]
How are you generating the flame graph? Leaks are normally detected using the `--leaks` argument that will show you the memory that has not been released at the end of the run. Read more about this here:

https://bloomberg.github.io/memray/flamegraph.html#memory-leaks-view

If you want to just profile a region of your code you can use the API:

https://bloomberg.github.io/memray/api.html

Alternatively you can use the temporal flame graphs with or without `--leaks` to explore those peaks. Learn more about this here:

https://bloomberg.github.io/memray/flamegraph.html#temporal-flame-graphs

**Reply by JackieJin1025** [reply]
Hi Pablogsal,

Thanks for your suggestions. I generated the flame graph without any additional arguments. I also tried generating the flame graph with the --temporal argument as you suggested. I've attached the result below.  
<img width="1778" alt="image" src="https://github.com/bloomberg/memray/assets/8197215/180e8c9f-39d7-43f1-beb6-1333106f1c33">

My initial question was that I have a loop that calls function simulation().  After each iteration, the memory usage grows.   My first impression was to find the memory difference between the two iterations. As the flame graph shows only the memory usage of simulation() once, I am not sure how to leverage the graph to detect the root cause. Generally speaking, how do I detect the memory usage over iterations if the flame graph shows the functions in iteration only once?

**Reply by pablogsal** [reply]
> As the flame graph shows only the memory usage of simulation() once, I am not sure how to leverage the graph to detect the root cause. Generally speaking, how do I detect the memory usage over iterations if the flame graph shows the functions in iteration only once?

I an not sure I fully understand why you say that "it shows the memory usage of simulation()" only once", but in general you can use the temporal flame graph with `--leaks` and then select a period in the time plot that you know it corresponds to a call to `simulation` and that will show you the memory that was allocated during the call that was not reallocated at the end, which corresponds to the memory "leaked" by a call to `simulation`.
